src/sfptpd_interface.c

```diff
@@ -2248,4 +2268,123 @@ int sfptpd_interface_driver_stats_reset(struct sfptpd_interface *interface)
 }
 
 
+int sfptpd_interface_reassign_to_nic(int from_phc, int to_phc)
+{
+	struct sfptpd_db_query_result q;
+	struct sfptpd_interface *intf;
+	struct sfptpd_interface *nic;
+	int my_false = 0;
+	int rc = 0;
+	int i;
+
+	interface_lock();
+
+	nic = FIND_ANY(INTF_KEY_CLOCK, &to_phc,
+		       INTF_KEY_DELETED, &my_false,
+		       SFPTPD_DB_SEL_END);
+	if (nic == NULL) {
+		ERROR("interface: could not find nic with phc %d\n", to_phc);
+		rc = ENODEV;
+		goto finish;
+	}
+
+	q = sfptpd_db_table_query(sfptpd_interface_table,
+				  INTF_KEY_CLOCK, &from_phc,
+				  INTF_KEY_DELETED, &my_false,
+				  SFPTPD_DB_SEL_END);
+
+	for (i = 0; i < q.num_records; i++) {
+		intf = *((struct sfptpd_interface **) q.record_ptrs[i]);
+		assert(intf->magic == SFPTPD_INTERFACE_MAGIC);
+		intf->nic_id = nic->nic_id;
+		intf->ts_info = nic->ts_info;
+		intf->clock = nic->clock;
+		intf->clock_supports_phc = nic->clock_supports_phc;
+		intf->driver_supports_efx = nic->driver_supports_efx;
+		INFO("interface: reassigned %s to nic id %d\n",
+		     intf->name, nic->nic_id);
+	}
+
+	q.free(&q);
+
+finish:
+	interface_unlock();
+	return rc;
+}
+
+
+int sfptpd_interface_hw_timestamping_suspend(struct sfptpd_db_query_result *q,
+					     int for_phc)
+{
+	struct hwtstamp_config req;
+	struct sfptpd_interface *intf;
+	int my_false = 0;
+	int rc = 0;
+	int i;
+
+	assert(q);
+	assert(for_phc >= 0);
+
+	interface_lock();
+
+	*q = sfptpd_db_table_query(sfptpd_interface_table,
+				   INTF_KEY_CLOCK, &for_phc,
+				   INTF_KEY_DELETED, &my_false,
+				   SFPTPD_DB_SEL_END);
+
+	for (i = 0; i < q->num_records; i++) {
+		intf = *((struct sfptpd_interface **) q->record_ptrs[i]);
+		assert(intf->magic == SFPTPD_INTERFACE_MAGIC);
+
+		rc = sfptpd_interface_ioctl(intf, SIOCGHWTSTAMP, &intf->saved_ts_conf);
+		if (rc != 0) {
+			ERROR("interface %s: could not get timestamping config for suspend, %s\n",
+			     intf->name, strerror(rc));
+			q->free(q);
+			goto fail;
+		}
+	}
+
+	for (i = 0; i < q->num_records; i++) {
+		intf = *((struct sfptpd_interface **) q->record_ptrs[i]);
+		memset(&req, 0, sizeof req);
+		req.tx_type = HWTSTAMP_TX_OFF;
+		req.rx_filter = HWTSTAMP_FILTER_NONE;
+		sfptpd_interface_ioctl(intf, SIOCSHWTSTAMP, &req);
+		INFO("interface %s: suspended hardware timestamping\n",
+		     intf->name);
+	}
```

Comment:
Looks like we don't free `q` on success, so the caller needs to do this. (Looks to be handled already providing there is a matching call to `sfptpd_interface_hw_timestamping_restore()`)

Reply:
Added comment to function.